presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaFilterManager.java

```diff
@@ -0,0 +1,319 @@
+
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.plugin.kafka.KafkaErrorCode.KAFKA_SPLIT_ERROR;
+import static io.prestosql.plugin.kafka.KafkaInternalFieldManager.OFFSET_TIMESTAMP_FIELD;
+import static io.prestosql.plugin.kafka.KafkaInternalFieldManager.PARTITION_ID_FIELD;
+import static io.prestosql.plugin.kafka.KafkaInternalFieldManager.PARTITION_OFFSET_FIELD;
+import static io.prestosql.spi.predicate.Marker.Bound.ABOVE;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaFilterManager
+{
+    private static final long INVALID_KAFKA_RANGE_INDEX = -1;
+    private static final String TOPIC_CONFIG_TIMESTAMP_KEY = "message.timestamp.type";
+    private static final String TOPIC_CONFIG_TIMESTAMP_VALUE_LOG_APPEND_TIME = "LogAppendTime";
+
+    private final KafkaConsumerFactory consumerFactory;
+    private final KafkaAdminFactory adminFactory;
+
+    @Inject
+    public KafkaFilterManager(KafkaConsumerFactory consumerFactory, KafkaAdminFactory adminFactory)
+    {
+        this.consumerFactory = requireNonNull(consumerFactory, "consumerManager is null");
+        this.adminFactory = requireNonNull(adminFactory, "adminFactory is null");
+    }
+
+    public KafkaFilteringResult getKafkaFilterResult(
+            ConnectorSession session,
+            KafkaTableHandle kafkaTableHandle,
+            List<PartitionInfo> partitionInfos,
+            Map<TopicPartition, Long> partitionBeginOffsets,
+            Map<TopicPartition, Long> partitionEndOffsets)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(kafkaTableHandle, "kafkaTableHandle is null");
+        requireNonNull(partitionInfos, "partitionInfos is null");
+        requireNonNull(partitionBeginOffsets, "partitionBeginOffsets is null");
+        requireNonNull(partitionEndOffsets, "partitionEndOffsets is null");
+
+        TupleDomain<ColumnHandle> constraint = kafkaTableHandle.getConstraint();
+        verify(!constraint.isNone(), "constraint is none");
+
+        if (!constraint.isAll()) {
+            Set<Long> partitionIds = partitionInfos.stream().map(partitionInfo -> (long) partitionInfo.partition()).collect(toImmutableSet());
+            Optional<Range> offsetRanged = Optional.empty();
+            Optional<Range> offsetTimestampRanged = Optional.empty();
+            Set<Long> partitionIdsFiltered = partitionIds;
+            Optional<Map<ColumnHandle, Domain>> domains = constraint.getDomains();
+
+            for (Map.Entry<ColumnHandle, Domain> entry : domains.get().entrySet()) {
+                KafkaColumnHandle columnHandle = (KafkaColumnHandle) entry.getKey();
```

Comment:
This loop feels wrong we are breaking whole process on the first non-internal column. 
And even if all constraints are on internal columns we are only processing arbitrary first one. 
Instead we should skip unsupported columns. And for internal once, gradually build up the predicate to be pushed down to kafka.